Add openmodelica header directory #1085

jbwahl opened the pull request:
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [x] Have you successfully run the Feel++ testsuite with your changes locally?

-----

prudhomm commented:
Could you please add some tests in the testsuite to use OMC ?

prudhomm commented on toolboxes/feel/feelmodels/hdg/poroelastic.hpp:
```diff
@@ -0,0 +1,379 @@
+	mpOptions.add ( makeMixedElasticityOptions("mixedelasticity") );
+
+	return mpOptions;
+}
```
function should go in a .cpp file. have just the function interface there

prudhomm commented on toolboxes/feel/feelmodels/hdg/poroelastic.hpp:
```diff
@@ -0,0 +1,379 @@
+}
+
+inline po::options_description
+makeMixedPoissonElasticityLibOptions( std::string prefix = "mixedpoissonelasticity" )
```
same as above

prudhomm commented on toolboxes/feel/feelmodels/hdg/poroelastic.hpp:
```diff
@@ -0,0 +1,379 @@
+	using Wh_tPOI = Pdh_type<mesh_type,Order>;
+
+	
+	using op_interp_ptrtypeEL = boost::shared_ptr<OperatorInterpolation<Wh_tEL, Pdhv_type<mesh_type,Order>>>;
```
indentation please ;)

prudhomm commented on toolboxes/feel/feelmodels/hdg/poroelastic.hpp:
```diff
@@ -0,0 +1,379 @@
+
+    mesh_ptrtype M_mesh;
+	
+	mp_ptrtype M_PoissonModel;
```
indentation

prudhomm commented on toolboxes/feel/feelmodels/hdg/poroelastic.hpp:
```diff
@@ -0,0 +1,379 @@
+#ifndef _POROELASTIC_HPP
```
please add common Feel++ header here

prudhomm commented on toolboxes/feel/feelmodels/hdg/poroelastic.hpp:
```diff
@@ -0,0 +1,379 @@
+
+template <int Dim, int Order,int G_Order, int E_Order>
+void
+MixedPoissonElasticity<Dim,Order,G_Order,E_Order>::assembleF_Poisson()
```
so _ in function names as per coding rules

prudhomm commented on toolboxes/hdg/poroelastic_main.cpp:
```diff
@@ -0,0 +1,118 @@
+#include "../feel/feelmodels/hdg/poroelastic.hpp"
```
add standard header
no ../ in include
use <feel/feelmodels/hdg/poroelastic.hpp>

jbwahl commented:
Implemented in this pull request : OpenModelica wrapper in Feelpp
The base class of any model is virtual and implemented in `feel/feelom/ommodel.hpp`
During the cmake installation of Feelpp, the declared OM models are generated as well as the specific header and library needed to wrap the model. 
The library is then generated using the proper make command.

Usage : 
- Considering a OM model described in some `.mo` or `.mos` files, in the `CMakelists.txt` of my application, you add the macros 
```
om_add_model( <NAME> SRCS <FILES> LIB <LIBNAME> )
```
with 
<NAME> : the name of the mode. IMPORTANT : it has to be the name of the main class/model in your OM model.
<FILES> : the `.mo` or `.mos` files which describe the OM model. IMPORTANT the main file has to be the first one declared. 
<LIBNAME> the name of the library to be generated. This library has then to be link to your application.

Example of declaration in `testsuite/feelom/CMakeLists.txt' : 
```
om_add_model( test_om SRCS test_om.mo LIB TEST_MO_LIB )
feelpp_add_application( test_om SRCS test_om.cpp LINK_LIBRARIES ${TEST_MO_LIB} )
```

To use the model, you have to add the header `"feel<NAME>.hpp" ' where <NAME> is the name declared in your `CMakeLists.txt`. 
You can then instantiate in your code the wrapper class and run the model with
```
Feel::Feel<NAME> my_om_model;
my_om_model.run();
```
The results of this run are written in the working directory, in the file `<NAME>.csv`

More features coming soon : 
- modification of the input variables through the wrapper
- acquisition of the output data through the wrapper.

prudhomm commented:
@jbwahl there is a conflict in this PR! we cannot merge currently

jbwahl commented:
@prudhomm right i noticed that.
I fixed it and i'm working on OM model compilation in cmake. I push it when i'm done.

jbwahl commented:
New features in the branch: 
- compile an executable from openmodelica model:
```
add_omc( <name> SRCS <srcs_files> )
```
will add the cmake target `feelpp_add_omc_<name>` which compile an executable from the <srcs_files>. The output is in `${CMAKE_CURRENT_BINARY_DIR}/<name>/`. 

WARNING: The first source file has to contain the main class of the model. Also, the <name> has to be the name of this main class!

- create an FMU file from an OpenModelica model:
```
add_fmu( <name> CLASS <class> SRCS <srcs_files> )
```
will add the cmake target `feelpp_add_fmu_<name>` which create an FMU file from OpenModelica `<srcs_files>`. Here the <name> can be set independently of the class name which specified in the <class> variable. The source files have to be ordered to put the main source file in the first position. 

The FMU file can now be loaded in Feelpp with limited features for now. See the test case in `testsuite/feelfmi/'  for an example.

@prudhomm @vincentchabannes I did not manage to do the "install" part of the macros 'add_fmu' and 'add_omc'. 
@prudhomm aside from this last point, I think we are good for a merge.

prudhomm commented:
We first need to check with buildkite that everything builds at least and also check the testsuite.
could you please rename the macro `feelpp_add_omc` and feelpp_add_fmu`

jbwahl commented:
@prudhomm should be ok now
